### PR TITLE
add support of text mode for services with only query

### PR DESCRIPTION
**Pull request by gs-gunjan**
## Summary
This change allows users to specify a service with only query. This support is only extended to text mode.

## How did you test this change?

- [x] Test(s) added
- [x] Manual testing (please provide screenshots/recordings)
- [ ] No testing (please provide an explanation)

### CODE REVIEW

**Comment by akphi** [comment]
@gs-gunjan I converted it draft, when this is read for review, please flip it back on. Thanks!

**Comment by gs-gunjan** [comment]
Screenshot (39)
Screenshot (40)
Screenshot (38)

**Comment by gs-gunjan** [comment]
legend engine change: https://github.com/finos/legend-engine/commit/818b0819181fbfe8aaf8924f3fb55b327f58de29

File: .changeset/breezy-pumas-own.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+---
+'@finos/legend-graph': major
+'@finos/legend-application-studio': minor
+'@finos/legend-query-builder': minor
+---
```
Comment (gayathrir11) [inline .changeset/breezy-pumas-own.md:5]: Can you please split the changeset. Just add an empty changeset for `legend-query-builder` with patch change

File: packages/legend-application-studio/src/components/editor/edit-panel/service-editor/ServiceExecutionQueryEditor.tsx [hunk 2]
```diff
@@ -327,11 +332,11 @@ export const ServiceExecutionQueryEditor = observer(
             executionState.setOpeningQueryEditor(false);
             return;
           }
+          applicationStore.notifyWarning(
+            'Please specify a mapping and a runtime for the execution context to edit with query builder',
+          );
+          executionState.setOpeningQueryEditor(false);
```
Comment (gayathrir11) [inline packages/legend-application-studio/src/components/editor/edit-panel/service-editor/ServiceExecutionQueryEditor.tsx:338]: Can you attach a screenshot showing you can open query builder with this change

**Comment by gs-gunjan** [comment]
Screenshot (44)

File: packages/legend-application-studio/src/stores/editor-state/element-editor-state/service/ServiceExecutionState.ts [hunk 3]
```diff
@@ -373,8 +373,8 @@ export class ServicePureExecutionQueryState extends LambdaEditorState {
 }
 
 export interface ServiceExecutionContext {
-  mapping: PackageableElementReference<Mapping>;
```
Comment (MauricioUyaguari) [inline packages/legend-application-studio/src/stores/editor-state/element-editor-state/service/ServiceExecutionState.ts:376]: The changes in this file make me a bit nervous as they could lead to some bugs in the future. 
Could we maybe add like a new `ServiceExecutionState` called `InlineServiceExecutionState` or something to that terms and just build that when we `buildExecutionState` if there are no mapping/runtime.
And then ensure `ServiceExecutionContext` always has a defined mapping/runtime. 
Wdyt?

File: packages/legend-application-studio/src/stores/shared/modifier/DSL_Service_GraphModifierHelper.ts [hunk 4]
```diff
@@ -202,7 +202,9 @@ export const pureSingleExecution_setMapping = action(
     value: Mapping,
     observerContext: ObserverContext,
   ): void => {
-    pe.mapping.value = observe_Mapping(value, observerContext);
+    if (pe.mapping !== undefined) {
```
Comment (MauricioUyaguari) [inline packages/legend-application-studio/src/stores/shared/modifier/DSL_Service_GraphModifierHelper.ts:205]: if(pe.mapping)

File: packages/legend-query-builder/src/stores/workflows/ServiceQueryBuilderState.ts [hunk 5]
```diff
@@ -77,7 +77,11 @@ export class ServiceQueryBuilderState extends QueryBuilderState {
     this.onServiceChange = onServiceChange;
     this.onExecutionContextChange = onExecutionContextChange;
 
-    if (service.execution instanceof PureSingleExecution) {
+    if (
+      service.execution instanceof PureSingleExecution &&
+      service.execution.mapping !== undefined &&
+      service.execution.runtime !== undefined
+    ) {
       this.mapping = service.execution.mapping.value;
```
Comment (MauricioUyaguari) [inline packages/legend-query-builder/src/stores/workflows/ServiceQueryBuilderState.ts:85]: lets instead add an asserTrue within this case statement
assertTrue(service.execution.mapping && service.execution.runtime, 'Service queries without runtime/mapping are not supported')

**Review by MauricioUyaguari — CHANGES_REQUESTED** [review]

File: packages/legend-manual-tests/src/__tests__/roundtrip-grammar/cases/DSL_Service-withOnlyQuery [hunk 6]
```diff
@@ -0,0 +1,184 @@
+###Data
```
[the remark this answers is not in the document]
Comment (gs-gunjan) [reply packages/legend-manual-tests/src/__tests__/roundtrip-grammar/cases/DSL_Service-withOnlyQuery:1]: actually we do not rely on .pure suffix, its only required while splitting and creating the test name, but that is optional, so as part of the run we do run this test, but will add the extension

File: packages/legend-graph/src/graphManager/protocol/pure/v1/transformation/pureGraph/to/helpers/V1_ServiceBuilderHelper.ts [hunk 7]
```diff
@@ -330,19 +330,34 @@ export const V1_buildServiceExecution = (
       serviceExecution.func,
       `Service Pure execution 'func' field is missing`,
     );
+    if (
+      serviceExecution.mapping !== undefined &&
+      serviceExecution.runtime !== undefined
+    ) {
```
[the remark this answers is not in the document]
Comment (gs-gunjan) [reply packages/legend-graph/src/graphManager/protocol/pure/v1/transformation/pureGraph/to/helpers/V1_ServiceBuilderHelper.ts:336]: if that is the case it would error out before doing anything with singleExec and throw an error

File: packages/legend-application-studio/src/components/editor/edit-panel/service-editor/ServiceEditor.tsx [hunk 8]
```diff
@@ -421,6 +427,9 @@ export const ServiceEditor = observer(() => {
       LEGEND_APPLICATION_DOCUMENTATION_KEY.QUESTION_HOW_TO_WRITE_A_SERVICE_TEST,
     );
 
+  const cannotOpenServiceTestEditor =
```
[the remark this answers is not in the document]
Comment (gs-gunjan) [reply packages/legend-application-studio/src/components/editor/edit-panel/service-editor/ServiceEditor.tsx]: done

File: packages/legend-application-studio/src/components/editor/edit-panel/service-editor/ServiceExecutionQueryEditor.tsx [hunk 9]
```diff
@@ -257,7 +257,12 @@ export const ServiceExecutionQueryEditor = observer(
         const service = executionState.serviceEditorState.service;
         const selectedExecutionState =
           executionState.selectedExecutionContextState;
-        if (selectedExecutionState) {
+        if (selectedExecutionState?.executionContext.mapping === undefined) {
+          applicationStore.notifyWarning(
```
Comment (MauricioUyaguari) [inline packages/legend-application-studio/src/components/editor/edit-panel/service-editor/ServiceExecutionQueryEditor.tsx:261]: notifyWarning -> notifyError

File: packages/legend-application-studio/src/stores/editor-state/element-editor-state/service/testable/ServiceTestDataState.ts [hunk 10]
```diff
@@ -366,7 +366,10 @@ export class ServiceTestDataState {
       this.testSuiteState.testableState.serviceEditorState.service;
     const execution = service.execution;
     let runtimes: Runtime[] = [];
-    if (execution instanceof PureSingleExecution) {
+    if (
+      execution instanceof PureSingleExecution &&
+      execution.runtime !== undefined
```
Comment (MauricioUyaguari) [inline packages/legend-application-studio/src/stores/editor-state/element-editor-state/service/testable/ServiceTestDataState.ts:371]: && execution.runtime

File: packages/legend-application-studio/src/components/editor/edit-panel/service-editor/ServiceExecutionEditor.tsx [hunk 11]
```diff
@@ -89,240 +89,245 @@ const PureExecutionContextConfigurationEditor = observer(
...
-    const mappingOptions =
-      editorStore.graphManagerState.usableMappings.map(buildElementOption);
-    const noMappingLabel = (
-      <div
-        className="service-execution-editor__configuration__mapping-option--empty"
-        title={isMappingEmpty?.messages.join('\n') ?? ''}
-      >
-        <div className="service-execution-editor__configuration__mapping-option--empty__label">
-          (none)
+    if (executionContext.mapping !== undefined) {
```
Comment (MauricioUyaguari) [inline packages/legend-application-studio/src/components/editor/edit-panel/service-editor/ServiceExecutionEditor.tsx:92]: if(executionContext.mapping)

File: packages/legend-application-studio/src/stores/editor-state/element-editor-state/service/ServiceExecutionState.ts [hunk 12]
```diff
@@ -394,13 +394,20 @@ export abstract class ServiceExecutionContextState {
 }
 
 export class SingleExecutionContextState extends ServiceExecutionContextState {
-  declare executionContext: PureSingleExecution;
+  declare executionContext1: PureSingleExecution;
```
Comment (MauricioUyaguari) [inline packages/legend-application-studio/src/stores/editor-state/element-editor-state/service/ServiceExecutionState.ts:397]: why was this renamed?

File: packages/legend-application-studio/src/stores/editor-state/element-editor-state/service/ServiceExecutionState.ts [hunk 13]
```diff
@@ -409,15 +416,18 @@ export class SingleExecutionContextState extends ServiceExecutionContextState {
   }
 
   setMapping(value: Mapping): void {
+    this.executionContext.mapping =
```
Comment (MauricioUyaguari) [inline packages/legend-application-studio/src/stores/editor-state/element-editor-state/service/ServiceExecutionState.ts:419]: can just do 
```
    pureSingleExecution_setMapping(
      this.executionContext1,
      PackageableElementExplicitReference.create(value),
      this.executionState.editorStore.changeDetectionState.observerContext,
    );

```

File: packages/legend-application-studio/src/stores/editor-state/element-editor-state/service/ServiceExecutionState.ts [hunk 14]
```diff
@@ -409,15 +416,18 @@ export class SingleExecutionContextState extends ServiceExecutionContextState {
...
+      PackageableElementExplicitReference.create(value);
     pureSingleExecution_setMapping(
-      this.executionContext,
+      this.executionContext1,
       value,
       this.executionState.editorStore.changeDetectionState.observerContext,
     );
   }
   setRuntime(value: Runtime): void {
+    this.executionContext.runtime = value;
```
Comment (MauricioUyaguari) [inline packages/legend-application-studio/src/stores/editor-state/element-editor-state/service/ServiceExecutionState.ts:428]: this line is not necessary, the `pureSingleExecution_setRuntime` sets it already

File: packages/legend-application-studio/src/stores/editor-state/element-editor-state/service/ServiceExecutionState.ts [hunk 15]
```diff
@@ -656,21 +678,19 @@ export abstract class ServicePureExecutionState extends ServiceExecutionState {
...
-        this.selectedExecutionContextState.executionContext.mapping.value,
-        this.editorStore,
-      );
-      this.selectedExecutionContextState.setRuntime(customRuntime);
-    }
+    const customRuntime = new EngineRuntime();
+    runtime_addMapping(
+      customRuntime,
+      PackageableElementExplicitReference.create(
+        this.selectedExecutionContextState!.executionContext.mapping.value,
```
Comment (MauricioUyaguari) [inline packages/legend-application-studio/src/stores/editor-state/element-editor-state/service/ServiceExecutionState.ts:685]: lets use guaranteeNonUllable for these two. 
guaranteeNonNullable(this.selectedExecutionContextState?.executionContext.mapping.value)

File: packages/legend-application-studio/src/stores/editor-state/element-editor-state/service/ServiceExecutionState.ts [hunk 16]
```diff
@@ -754,23 +816,28 @@ export class SingleServicePureExecutionState extends ServicePureExecutionState {
...
-    );
-    _execution.executionParameters = [_parameter];
-    service_setExecution(
-      this.serviceEditorState.service,
-      _execution,
-      this.editorStore.changeDetectionState.observerContext,
-    );
-    this.serviceEditorState.resetExecutionState();
+    if (
+      this.execution.mapping !== undefined &&
```
Comment (MauricioUyaguari) [inline packages/legend-application-studio/src/stores/editor-state/element-editor-state/service/ServiceExecutionState.ts:820]: this.execution.mapping && this.execution.runtime

**Review by MauricioUyaguari — CHANGES_REQUESTED** [review]
small comments

File: packages/legend-query-builder/src/stores/workflows/ServiceQueryBuilderState.ts [hunk 17]
```diff
@@ -78,7 +78,12 @@ export class ServiceQueryBuilderState extends QueryBuilderState {
     this.onExecutionContextChange = onExecutionContextChange;
 
     if (service.execution instanceof PureSingleExecution) {
-      this.mapping = service.execution.mapping.value;
+      assertTrue(
+        service.execution.mapping !== undefined &&
```
[the remark this answers is not in the document]
Comment (gs-gunjan) [reply packages/legend-query-builder/src/stores/workflows/ServiceQueryBuilderState.ts:82]: For this particular change it was giving me an error, I was able to change the others

File: packages/legend-graph/src/graphManager/protocol/pure/v1/transformation/pureGraph/from/V1_ServiceTransformer.ts [hunk 18]
```diff
@@ -136,8 +136,10 @@ const transformSingleExecution = (
   execution.func = element.func.accept_RawValueSpecificationVisitor(
     new V1_RawValueSpecificationTransformer(context),
   ) as V1_RawLambda;
-  execution.mapping = element.mapping.valueForSerialization ?? '';
-  execution.runtime = V1_transformRuntime(element.runtime, context);
+  if (element.mapping && element.runtime) {
```
[the remark this answers is not in the document]
Comment (gs-gunjan) [reply packages/legend-graph/src/graphManager/protocol/pure/v1/transformation/pureGraph/from/V1_ServiceTransformer.ts:139]: the execution will be built only if both are present

File: packages/legend-application-studio/src/stores/editor-state/element-editor-state/service/testable/ServiceTestDataState.ts (hunk 10 above)
Comment (gs-gunjan) [reply packages/legend-application-studio/src/stores/editor-state/element-editor-state/service/testable/ServiceTestDataState.ts:371]: tried the suggestion out was giving a lint error, didnt dvelve much into why

File: packages/legend-application-studio/src/stores/editor-state/element-editor-state/service/ServiceExecutionState.ts (hunk 16 above)
Comment (gs-gunjan) [reply packages/legend-application-studio/src/stores/editor-state/element-editor-state/service/ServiceExecutionState.ts:820]: same problem here

File: packages/legend-application-studio/src/stores/editor-state/element-editor-state/service/ServiceExecutionState.ts (hunk 12 above)
Comment (gs-gunjan) [reply packages/legend-application-studio/src/stores/editor-state/element-editor-state/service/ServiceExecutionState.ts:397]: So this property comes from parent class and because now we change given mapping and runtime could be defined, so couldnt have the same name

File: packages/legend-application-studio/src/stores/editor-state/element-editor-state/service/ServiceExecutionState.ts (hunk 13 above)
Comment (gs-gunjan) [reply packages/legend-application-studio/src/stores/editor-state/element-editor-state/service/ServiceExecutionState.ts:419]: I had to explicit set it, otherwise the runtime value was not getting populated readily in the UI

File: packages/legend-application-studio/src/stores/editor-state/element-editor-state/service/ServiceExecutionState.ts (hunk 15 above)
Comment (gs-gunjan) [reply packages/legend-application-studio/src/stores/editor-state/element-editor-state/service/ServiceExecutionState.ts:685]: same her as well

File: packages/legend-application-studio/src/stores/editor-state/element-editor-state/service/ServiceExecutionState.ts (hunk 14 above)
Comment (gs-gunjan) [reply packages/legend-application-studio/src/stores/editor-state/element-editor-state/service/ServiceExecutionState.ts:428]: same here as well

File: packages/legend-graph/src/graphManager/protocol/pure/v1/transformation/pureGraph/from/V1_ServiceTransformer.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (MauricioUyaguari) [reply packages/legend-graph/src/graphManager/protocol/pure/v1/transformation/pureGraph/from/V1_ServiceTransformer.ts]: yes but thats the not job of the transformer. the job of the transformer to just transform the metamodel to protocol. The builder is the one who checks or has that logic.

**Comment by MauricioUyaguari** [comment]
@gs-gunjan could we fix the code checks for the build. Also can we rebase (this will clear up the intergration tests that are failing). After that I can merge. I'll address at the other comments I left. 
Thankis!

**Review by MauricioUyaguari — APPROVED** [review]
